mne/stats/cluster_level.py

```diff
@@ -1729,3 +1737,310 @@ def summarize_clusters_stc(
+        y, X = formulaic.model_matrix(formula, pivot_df)
+    else:
+        raise ValueError(
+            "Formula is required and needs to be a string in Wilkinson notation."
+        )
+
+    # now prep design matrix for input into MNE cluster function
+    # cluster functions expects channels as list dimension
+    y_for_cluster = y.values.reshape(-1, n_channels, n_timepoints).transpose(0, 2, 1)
+
+    adjacency, _ = find_ch_adjacency(df["evoked"][0].info, ch_type="eeg")
+
+    # define stat function and threshold
+    stat_fun, threshold = _check_fun(
+        X=y_for_cluster, stat_fun=None, threshold=None, tail=0, kind="within"
+    )
+
+    # Run the cluster-based permutation test
+    T_obs, clusters, cluster_p_values, H0 = _permutation_cluster_test(
+        [y_for_cluster],
+        n_permutations=n_permutations,
+        threshold=threshold,
+        stat_fun=stat_fun,
+        tail=tail,
+        n_jobs=n_jobs,
+        adjacency=adjacency,
+        max_step=max_step,  # maximum distance between samples (time points)
+        exclude=exclude,  # exclude no time points or channels
+        step_down_p=step_down_p,  # step down in jumps test
+        t_power=t_power,  # weigh each location by its stats score
+        out_type=out_type,
+        check_disjoint=check_disjoint,
+        buffer_size=buffer_size,  # block size for chunking the data
+        seed=seed,
+    )
+
+    print(f"smallest cluster p-value: {min(cluster_p_values)}")
+
+    return ClusterResult(T_obs, clusters, cluster_p_values, H0)
+
+
+def unpack_time_and_channels(df: pd.DataFrame = None) -> pd.DataFrame:
+    """
+    Extract timepoints and channels and convert to long.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        DataFrame in wide format.
+
+    Returns
+    -------
+    df_long : pd.DataFrame
+        DataFrame in long format.
+    """
+    # Extracting all necessary data using list comprehensions for better performance
+    long_format_data = [
+        {
+            "condition": row["condition"],
+            "subject_index": row["subject_index"],
+            "channel": channel,
+            "timepoint": timepoint,
+            "value": row["data"][channel, timepoint],
+        }
+        for idx, row in df.iterrows()
+        for channel in range(row["data"].shape[0])
+        for timepoint in range(row["data"].shape[1])
+    ]
+
+    # Creating the long format DataFrame
+    df_long = pd.DataFrame(long_format_data)
+
+    return df_long
+
+
+class ClusterResult:
+    """
+    Object containing the results of the cluster permutation test.
+
+    Parameters
+    ----------
+    T_obs : np.ndarray
+        The observed test statistic.
+    clusters : list
+        List of clusters.
+    cluster_p_values : np.ndarray
+        P-values for each cluster.
+    H0 : np.ndarray
+        Max cluster level stats observed under permutation.
+    """
+
+    def __init__(self, T_obs, clusters, cluster_p_values, H0):
```

Comment:
type annotations are still missing here 

for arrays, consider using numpy.typing.NDarray